### PR TITLE
Config schema

### PR DESCRIPTION
**Pull request by TheFrok**
Added JSON Schema validation for `config.json`
and used the schema to generate sphinx doc using `sphinx-jsonschema`
Issue #266

### CODE REVIEW

**Comment by TheFrok** [comment]
@OrBin the tests fail because I deleted the option to use future_days_count, and I don't know how to change that on the Travis machine config file.

**Comment by OrBin** [comment]
> @OrBin the tests fail because I deleted the option to use future_days_count, and I don't know how to change that on the Travis machine config file.

As you can see in [.travis.yml](https://github.com/hasadna/open-bus/blob/fc9bf7b2d2401a51ea7180b23ccf0527a11cc070/.travis.yml#L12), the configuration is taken from `gtfs/gtfs_utils/gtfs_utils/config_example.json`. If you removed `future_days_count`, do so in the configuration example too.

File: gtfs/gtfs_utils/gtfs_utils/config.schema.json [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    "type": "object",
+    "properties": {
+      "files": {
+        "$ref": "#/definitions/files"
+      },
+      "s3": {
+        "$ref": "#/definitions/s3"
+      },
+      "date_range": {
+        "description": "First and last date of the gtfs files to be download from bucket. {Format: *YYYY-MM-DD*}",
```
Comment (OrBin) [inline gtfs/gtfs_utils/gtfs_utils/config.schema.json]: Add a note that this field is only relevant if `use_data_from_today` is `false`.

File: gtfs/gtfs_utils/gtfs_utils/config.schema.json [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        "type": "array",
+        "items": {
+          "type": "string",
+          "pattern": "^\\d{4}-\\d{2}-\\d{2}$"
+        },
+        "maxLength": 2,
+        "minLength": 2
+      },
+      "use_data_from_today": {
+        "description": "Download only gtfs data from today (overrides date_range). {default: *False*}",
```
Comment (OrBin) [inline gtfs/gtfs_utils/gtfs_utils/config.schema.json]: Use `default` in the schema instead of writing it in the description. Also, it's JSON, so `false`, with lowercase f. Sorry for the pettiness :)

File: gtfs/gtfs_utils/gtfs_utils/config.schema.json [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        "type": "boolean"
+      },
+      "max_gtfs_size_in_mb": {
+        "description": "Limit the maximum size of the downloaded gtfs files (in MB). If not set the limit is only free disk space.",
```
Comment (OrBin) [inline gtfs/gtfs_utils/gtfs_utils/config.schema.json]: Please add a comma after "If not set". The same for "If true, " and "If false, " in the following fields.

File: gtfs/gtfs_utils/gtfs_utils/config.schema.json [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        "description": "If true writes a filtered version of the gtfs for the specific dates. {default: *False*}",
+        "type": "boolean"
+      },
+      "console_verbosity": {
+        "description": "Lowest logging level to be printed to console. {default: *'Error'*}",
+        "type": "string"
+      }
+    },
+    "required": [
+      "date_range",
```
Comment (OrBin) [inline gtfs/gtfs_utils/gtfs_utils/config.schema.json]: Are you sure it should be required? I'm not sure how such cases should be handled.

File: gtfs/gtfs_utils/gtfs_utils/config.schema.json [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+  "$comment": "The definitions are separated to a different key so it won't appear in the main doc",
+  "definitions": {
+    "files": {
+      "title": "Files object parameters",
+      "$$target": "#/definitions/files",
+      "description": "files info would be under the `files` tag.",
+      "type": "object",
+      "properties": {
+        "base_directory": {
+          "description": "base directory for the for the created files.",
```
Comment (OrBin) [inline gtfs/gtfs_utils/gtfs_utils/config.schema.json]: "for the for the" -> "for the"

File: gtfs/gtfs_utils/gtfs_utils/config.schema.json [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+          "type": "string"
+        },
+        "output_file_name_regexp": {
+          "description": "A regular expression used to find existing output files.",
+          "type": "string"
+        },
+        "output_file_type": {
+          "description": "The file type for the outputs (for example, csv.gz).",
```
Comment (OrBin) [inline gtfs/gtfs_utils/gtfs_utils/config.schema.json]: Please list the available types.

File: gtfs/gtfs_utils/gtfs_utils/config.schema.json [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+          "type": "string"
+        },
+        "child_directories": {
+          "type": "object",
+          "properties": {
+            "gtfs_feeds": {
+              "description": "The name of the directory for the gtfs_feeds.",
```
Comment (OrBin) [inline gtfs/gtfs_utils/gtfs_utils/config.schema.json]: "gtfs_feeds" -> "gtfs feeds"
Also, it'd be better to note that this is the directory the GTFS files **will be downloaded to**.

File: gtfs/gtfs_utils/gtfs_utils/config.schema.json [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+      "$$target": "#/definitions/s3",
+      "description": "All the info about the s3 connection parameters would come here (under `s3` tag).",
+      "type": "object",
+      "properties": {
+        "access_key_id": {
+          "description": "Authorization access key id for S3.",
+          "type": "string"
+        },
+        "secret_access_key": {
+          "description": "Authorization secert access key for S3.",
```
Comment (OrBin) [inline gtfs/gtfs_utils/gtfs_utils/config.schema.json]: secert -> secret

File: gtfs/gtfs_utils/gtfs_utils/config.schema.json [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+          "type": "string"
+        },
+        "s3_endpoint_url": {
+          "description": "Connection endpoint url for S3.",
+          "type": "string"
+        },
+        "bucket_name": {
+          "description": "Connection bucket name for S3.",
```
Comment (OrBin) [inline gtfs/gtfs_utils/gtfs_utils/config.schema.json]: "Connection bucket name" -> "Bucket name"

File: gtfs/gtfs_utils/gtfs_utils/config.schema.json [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+          "type": "string"
+        },
+        "upload_results": {
+          "description": "If true upload the analzed results back to the S3.",
```
Comment (OrBin) [inline gtfs/gtfs_utils/gtfs_utils/config.schema.json]: analzed -> analyzed
**the** S3 -> S3

File: gtfs/gtfs_utils/gtfs_utils/config.schema.json [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+          "type": "boolean"
+        },
+        "results_path_prefix": {
+          "description": "Prefix for the uploaded results.",
```
Comment (OrBin) [inline gtfs/gtfs_utils/gtfs_utils/config.schema.json]: Note that this is the prefix **for the path in S3**.

File: gtfs/docs/source/conf.py [hunk 12]
```diff
@@ -93,6 +94,12 @@
 # so a file named "default.css" will overwrite the builtin "default.css".
 html_static_path = ['_static']
 
+
+# Known issue with line wrap inside table for ReadTheDocs:
+# https://github.com/readthedocs/sphinx_rtd_theme/issues/117
+def setup(app):
```
[the remark this answers is not in the document]
Comment (TheFrok) [reply gtfs/docs/source/conf.py]: It called from sphinx internal. 
declaring a `setup` function lets you treat the `conf.py` as a sphinx extension to alter the rendering.
https://www.sphinx-doc.org/en/master/extdev/appapi.html

File: gtfs/gtfs_utils/gtfs_utils/configuration.py [hunk 13]
```diff
@@ -109,9 +111,26 @@ def dict_to_dataclass(data_dict: Dict, data_class: type) -> Dict:
...
+def validate_configuration_schema(configuration_dict: dict) -> bool:
+    """
+    Validate the configuration dict against the configuration JSON Schema
+    :param configuration_dict: a dict of the configuration, as loaded with json.load
+    :return: True or raises an error (jsonschema.exceptions.ValidationError)
+    """
+    with open(CONFIGURATION_SCHEMA_FILE_PATH, 'r') as schema_file:
+        config_schema = json.load(schema_file)
+        # The definitions are separated to a different key,
+        # so it won't appear in the sphinx generated docs
```
[the remark this answers is not in the document]
Comment (TheFrok) [reply gtfs/gtfs_utils/gtfs_utils/configuration.py]: Not really, I'm doing that just because I want to use the sphinx extension and I wan to control the documentation. Another option would be to separate the definitions into a different file, but I thought it would be to complicated.
That's the reason I must rejoin the json for the validation to work.

You can see my conversation with the creator of the package here -
https://github.com/lnoor/sphinx-jsonschema/issues/23

File: gtfs/gtfs_utils/gtfs_utils/config.schema.json (hunk 4 above)
Comment (TheFrok) [reply gtfs/gtfs_utils/gtfs_utils/config.schema.json]: I can set it as only one of `use_data_from_today` or `date_range` would be required. It's a bit ugly with the date_range defult value:
`date_range: Iterable[str] = ('',)`

File: gtfs/gtfs_utils/gtfs_utils/configuration.py [hunk 14]
```diff
@@ -70,7 +70,7 @@ class Configuration:
     files: FilesConfiguration = None
     s3: S3Configuration = None
     use_data_from_today: bool = True
-    date_range: List[str] = None
+    date_range: Iterable[str] = ('',)
```
Comment (OrBin) [inline gtfs/gtfs_utils/gtfs_utils/configuration.py]: Why did you change to iterable?

File: gtfs/gtfs_utils/gtfs_utils/config.schema.json [hunk 15]
```diff
@@ -67,22 +79,23 @@
       "type": "object",
       "properties": {
         "base_directory": {
-          "description": "base directory for the for the created files.",
+          "description": "base directory the for the created files.",
```
Comment (OrBin) [inline gtfs/gtfs_utils/gtfs_utils/config.schema.json]: Still one redundant "the" :)

File: gtfs/gtfs_utils/gtfs_utils/config.schema.json [hunk 16]
```diff
@@ -115,23 +128,23 @@
...
           "type": "string"
         },
         "upload_results": {
-          "description": "If true upload the analzed results back to the S3.",
+          "description": "If true, upload the analyzed results back to S3.",
           "type": "boolean"
         },
         "results_path_prefix": {
-          "description": "Prefix for the uploaded results.",
+          "description": "Prefix path on the S3 for the uploaded results.",
```
Comment (OrBin) [inline gtfs/gtfs_utils/gtfs_utils/config.schema.json]: the S3 -> S3

File: gtfs/gtfs_utils/gtfs_utils/config.schema.json (hunk 4 above)
Comment (OrBin) [reply gtfs/gtfs_utils/gtfs_utils/config.schema.json]: Why is this related to the default `date_range`?

File: gtfs/gtfs_utils/gtfs_utils/configuration.py (hunk 14 above)
Comment (TheFrok) [reply gtfs/gtfs_utils/gtfs_utils/configuration.py]: For it to not be a required field (if some one wants just "use_today_data"), it must have a default value.
Since it is impossible to set a mutable as default values in `dataclass` (only with a default factory) I used tuple.
because there is a type check in `dict_to_dataclass` the default value must be an instance of the hinted type.

File: gtfs/gtfs_utils/gtfs_utils/config.schema.json (hunk 4 above)
Comment (TheFrok) [reply gtfs/gtfs_utils/gtfs_utils/config.schema.json]: I thought you wanted `date_range` to not be required. I explained it in more depth in your other comment.

File: gtfs/gtfs_utils/gtfs_utils/configuration.py (hunk 14 above)
Comment (OrBin) [reply gtfs/gtfs_utils/gtfs_utils/configuration.py]: And what about using `dataclasses.field(default_factory=list)` as a default value?
It just doesn't feel right to change all this only for making it non-required. Doesn't JSON schema have an option to set an optional field?

File: gtfs/gtfs_utils/gtfs_utils/config.schema.json (hunk 4 above)
Comment (OrBin) [reply gtfs/gtfs_utils/gtfs_utils/config.schema.json]: OK got it. I'm resolving this comment and moving the discussion to the other one.

File: gtfs/gtfs_utils/gtfs_utils/configuration.py (hunk 14 above)
Comment (TheFrok) [reply gtfs/gtfs_utils/gtfs_utils/configuration.py]: Ok I found why I had a problem with the default_factory, will commit soon.
BTW, I noticed [this commit](https://github.com/hasadna/open-bus/commit/821e3b5d111951a84a3e1bf9715f55eda9864c3b) where you allowed configuration keys that are not in the data class. right now the JSON Schema does not allow that. I think that if someone adds a new key it should be added to the schema and docs as well.

Comment (OrBin) [reply gtfs/gtfs_utils/gtfs_utils/configuration.py]: Agreed.

**Comment by TheFrok** [comment]
```
[ERROR] Failed to execute goal org.springframework.boot:spring-boot-maven-plugin:2.0.2.RELEASE:repackage (default) on project bus: Execution default of goal org.springframework.boot:spring-boot-maven-plugin:2.0.2.RELEASE:repackage failed: Plugin org.springframework.boot:spring-boot-maven-plugin:2.0.2.RELEASE or one of its dependencies could not be resolved: Could not transfer artifact org.apache.commons:commons-compress:jar:1.14 from/to google-maven-central (https://maven-central.storage-download.googleapis.com/maven2/): GET request of: org/apache/commons/commons-compress/1.14/commons-compress-1.14.jar from google-maven-central failed: Connection reset -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginResolutionException
```

I don't think that's related to my change

File: gtfs/gtfs_utils/gtfs_utils/configuration.py [hunk 17]
```diff
@@ -77,15 +79,15 @@ class Configuration:
...
+def dict_to_dataclass(data_dict: Dict, data_class: type) -> Configuration:
     """
     Converts the dict to a dataclass instance of the given type.
     """
     data_class_instance = data_class()
 
     for field in fields(data_class):
         if field.name not in data_dict:
-            value = field.default
+            continue  # The value is already set by default value of the class
```
Comment (OrBin) [inline gtfs/gtfs_utils/gtfs_utils/configuration.py]: Are you sure that's OK? did you test it?

**Comment by OrBin** [comment]
> ```
> [ERROR] Failed to execute goal org.springframework.boot:spring-boot-maven-plugin:2.0.2.RELEASE:repackage (default) on project bus: Execution default of goal org.springframework.boot:spring-boot-maven-plugin:2.0.2.RELEASE:repackage failed: Plugin org.springframework.boot:spring-boot-maven-plugin:2.0.2.RELEASE or one of its dependencies could not be resolved: Could not transfer artifact org.apache.commons:commons-compress:jar:1.14 from/to google-maven-central (https://maven-central.storage-download.googleapis.com/maven2/): GET request of: org/apache/commons/commons-compress/1.14/commons-compress-1.14.jar from google-maven-central failed: Connection reset -> [Help 1]
> [ERROR] 
> [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
> [ERROR] Re-run Maven using the -X switch to enable full debug logging.
> [ERROR] 
> [ERROR] For more information about the errors and possible solutions, please read the following articles:
> [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginResolutionException
> ```
> 
> I don't think that's related to my change

You are right, it's in the Java section of the CI. I'll try to look into this/restart the job.

File: gtfs/gtfs_utils/gtfs_utils/configuration.py (hunk 17 above)
Comment (TheFrok) [reply gtfs/gtfs_utils/gtfs_utils/configuration.py]: Yes, The default values are set when we initialize the instance (`data_class_instance = data_class()`).
The only difference now is that it skips the type check when the default value is used, I thought it's redundant to validate the type for default values but I can change that.

Comment (OrBin) [reply gtfs/gtfs_utils/gtfs_utils/configuration.py]: I agree, just making sure it's tested :)

Comment (OrBin) [reply gtfs/gtfs_utils/gtfs_utils/configuration.py]: Actually, if it's OK this way, you can even change it to a positive condition:
Instead of using 
```python
if field.name not in data_dict:
	continue
else:
	...
```

you can simply change it to:
```python
if field.name in data_dict:
	...
```

**Comment by OrBin** [comment]
Closed #266
